apps/ogr2ogr_lib.cpp

```diff
@@ -4400,6 +4401,48 @@ std::unique_ptr<TargetLayerInfo> SetupTargetLayer::Setup(OGRLayer* poSrcLayer,
         psInfo->m_oMapDomainToKV[poDomain] = std::move(oMapCodeValue);
     }
 
+    // Detect if we can directly passed the source feature to the CreateFeature()
+    // method of the target layer, without doing any copying of field content.
+    psInfo->m_bCanAvoidSetFrom = false;
+    if( !m_bExplodeCollections && iSrcZField == -1 )
+    {
+        psInfo->m_bCanAvoidSetFrom = true;
+        const int nDstGeomFieldCount = poDstFDefn->GetGeomFieldCount();
+        if( nSrcFieldCount != poDstFDefn->GetFieldCount() ||
+            nSrcGeomFieldCount != nDstGeomFieldCount )
+        {
+            psInfo->m_bCanAvoidSetFrom = false;
+        }
+        else
+        {
+            for( int i = 0; i < nSrcFieldCount; ++i )
+            {
+                auto poSrcFieldDefn = poSrcFDefn->GetFieldDefn(i);
+                auto poDstFieldDefn = poDstFDefn->GetFieldDefn(i);
+                if( poSrcFieldDefn->GetType() != poDstFieldDefn->GetType() ||
+                    psInfo->m_anMap[i] != i )
```

Comment:
Do we need to consider field truncation here if the destination field has a size limit smaller than the source? (Or will that be handled automatically elsewhere at the driver level?)

Reply:
This consideration is not even taken into account by the non-optimized OGRFeature::SetFrom() method. At the OGRFieldDefn level, field width semantics is also probably a bit fuzzy currently. Some drivers might interpret it as a number of bytes and others as a number of UTF-8 characters. Some drivers can also happily store a string that is larger the indicated field size.